src/blocking/spi.rs

```diff
@@ -18,6 +36,16 @@ pub trait Write<W> {
     fn write(&mut self, words: &[W]) -> Result<(), Self::Error>;
 }
 
+/// Blocking read
+pub trait Read<W> {
```

Comment:
previously we elected to not offer a `read` type as the behavior of the data output would be undefined (or dependent on your hardware) which could lead to surprising behavior on real peripherals, where a `transfer` with zero's (or other value) in the output buffer is more explicit.

is there a new / improved argument for adding this?

https://github.com/rust-embedded/embedded-hal/issues/52

Reply:
:sweat_smile: I hadn't seen that issue.

It's true that leaving it implementation-defined limits the usefulness for hal-independent drivers. Maybe it could be specified to always send 0?

The usefulness I see is:
- No need to zero big buffers if we just want to read big data (can use an buffer with whatever data without leaking the data on MOSI)
- If we do add a way to read to uninit buffers (#288) then Read would be needed
- When using an SPI with only MISO pin, no MOSI pin.

Maybe it's better to leave it for later.

Reply:
> It's true that leaving it implementation-defined limits the usefulness for hal-independent drivers. Maybe it could be specified to always send 0?

so long as implementations support it, [linux](https://github.com/spotify/linux/blob/master/include/linux/spi/spidev.h#L54) appears to be okay with shifting zeros.

> No need to zero big buffers if we just want to read big data

yep, i suppose there's a use case for this with SPI flash.

> If we do add a way to read to uninit buffers (Methods to read into a buffer should take an uninitialized buffer #288) then Read would be needed

won't `transfer` also take a `MaybeUninit` read buffer in this case?

> When using an SPI with only MISO pin, no MOSI pin

is this an thing one would actually do? :-P


i'm alright with adding a `Read` (trait and `Transational` variant) if it's clearly documented that the write-data will be zeroed, @rust-embedded/hal any other opinions?

Reply:
> won't transfer also take a MaybeUninit read buffer in this case?

Single-buf Transfer would take a `&mut [u8]`. Multi-buf Transfer would take `read: &mut [MaybeUninit<u8>], write: &[u8]`. Still not useful to read a big buf without having to zero-out a big write buf, only `Read` can do that.

> is this an thing one would actually do? :-P

People do weird hacks with readonly SPI. For example a fancy logic analyzer that bit-packs the samples for you... however in that case the user will be OK with using the HAL directly or even the PAC, so this is a weak argument for adding it to embedded-hal, yup.

Reply:
Could we just say the write data will be undefined, and implementations might use 0, 1, or whatever? If you need the write data to be something particular you can use the transfer method and specify it; if you're just reading and don't care what's written then it should be OK to let the implementation decide whatever's best. I guess we can also provide a default impl of read in terms of transfer.

If you were in a MISO-only situation then the guarantee wouldn't even make sense since MOSI wouldn't have a defined level, and in any event I think it's typical that if you're reading you really don't care what's being written. The same is true of half-duplex buses using one wire for combined MISO and MOSI, where the master won't drive the output at all while reading - so again there's no defined level.

Still, I guess implementations have to pick something and I can't think of any practical situation where it's easier for the implementation to use 1 instead of 0 but in such a way that the actual application doesn't care. So maybe it's less confusing to just require writing all 0s.

Reply:
on nRF it's configurable. I'd prefer specifying it as "implementation defined" so it leaves room for HALs to add a method to configure it.